### PR TITLE
Intercept all Pulsar command and REST API requests.

### PR DESCRIPTION
**Pull request by codelipenghui**
### Motivation

Broker Interceptors

### Modifications

*Describe the modifications you've done.*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by jerrypeng** [comment]
@codelipenghui thanks for working on something like this.  The funny thing is we have developed an "interceptor" framework as well.  It covers all the REST endpoints as well.  Maybe we can collaborate?

**Comment by sijie** [comment]
@jerrypeng do you want to send out the pull request? so we can merge efforts here.

**Comment by merlimat** [comment]
Added PR: #7159

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java [hunk 1]
```diff
@@ -1769,6 +1783,20 @@
     )
     private Set<String> brokerClientTlsProtocols = Sets.newTreeSet();
 
+    /**** --- Interceptor variables --- ****/
+
+    @FieldContext(
+        category = CATEGORY_INTERCEPTORS,
+        doc = "The directory to locate interceptors"
+    )
+    private String interceptorDirectory = "./interceptors";
```
Comment (sijie) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:1792]: overall looks good to me. I see there are two sets of settings here, one is for `listeners` and the other one is for `interceptors`? I think we should just rename the listener interfaces to Interceptors. So it is consistent with client-side interceptor.

Comment (codelipenghui) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:1792]: Ok.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java [hunk 2]
```diff
@@ -450,7 +453,9 @@ public void start() throws PulsarServerException {
 
             this.defaultOffloader = createManagedLedgerOffloader(
                     OffloadPolicies.create(this.getConfiguration().getProperties()));
-
+            this.brokerInterceptor = BrokerInterceptors.load(config);
+            brokerService.setInterceptor(getBrokerInterceptor());
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java:457]: We should support multiple broker interceptors.  Can you add a broker interceptor implementation that wraps a list of interceptors into one interceptor?

Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java:457]: The current implementation is supporting multiple broker interceptors, I added BrokerInterceptors.java also implement BrokerInterceptor interface.

**Comment by jerrypeng** [comment]
@merlimat put up our version of this. I thought we were going to work together merge both our implementation?

**Comment by merlimat** [comment]
There are many points that are missing in this implementation. Especially on how to give a proper error to the client when some operation is denied.

**Comment by merlimat** [comment]
Also, this is exposing various internal classes to the interceptor API. That means that these classes are now becoming part of public API, which is not a good idea since they were not meant to be.

**Comment by sijie** [comment]
@merlimat @jerrypeng I think this is to set up the framework so the broker can load interceptors from NAR packages. The interface is very basic and only focuses on intercepting connections. We don't expect this interface to the public at this moment. It is the first step to implement interceptors. We should iterate the interface and add the pull request #7159 that @merlimat sent for intercepting per restful call.

Since these two pull requests have different focuses (NAR class loading, connection level interception vs per-rest-call interception) and they can be merged together, I don't see we need to block one pull request for the other pull request. Especially in any case, we need to load the plugin from NAR package just as what we did for other plugins, correct?

**Pull request merged**